src/index.ts

```diff
@@ -0,0 +1,87 @@
+export type StackErrorJSON = {
+  code: string
+  message: string
+  metadata: Record<string, unknown>
+  name: string
+  stack: Array<StackErrorJSON>
+}
+
+export class StackError extends Error {
+  static fromJSON(json: StackErrorJSON): StackError {
+    const error = new StackError(json.code, json.message)
+    error.errorStack = (json.stack ?? []).reduceRight((stack, e) => {
+      const err = StackError.fromJSON(e)
+      err.errorStack = stack
+      return [err, ...stack]
+    }, [] as Array<StackError>)
+    error.metadata = json.metadata ?? {}
+    error.name = json.name ?? 'StackError'
+    return error
+  }
+
+  code: string
+  errorStack: Array<StackError>
+  metadata: Record<string, unknown> = {}
+  name = 'StackError'
+
+  constructor(code: string, message: string, parentError?: StackError) {
+    super(message)
+    Object.setPrototypeOf(this, StackError.prototype)
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, StackError)
+    }
+
+    this.code = code
+    this.errorStack = parentError ? parentError.toErrorStack() : []
+  }
+
+  toErrorStack(): Array<StackError> {
+    return [this, ...this.errorStack]
+  }
+
+  toJSON(withStack = true): StackErrorJSON {
+    return {
+      code: this.code,
+      message: this.message,
+      metadata: this.metadata,
+      name: this.name,
+      stack: withStack ? this.errorStack.map(e => e.toJSON(false)) : []
+    }
+  }
+
+  toString(): string {
+    return `[${this.code}] ${this.message}`
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function createNamespaceError(namespace: string, metadata: Record<string, unknown> = {}) {
+  return class NamespaceError extends StackError {
+    constructor(code: string | number, message: string, parentError?: StackError) {
+      super(`${namespace}${code}`, message, parentError)
+      Object.setPrototypeOf(this, NamespaceError.prototype)
+      if (Error.captureStackTrace) {
+        Error.captureStackTrace(this, NamespaceError)
+      }
+
+      this.metadata = metadata
+    }
+  }
+}
+
+export function assert(condition: boolean, message = 'Assertion failed'): asserts condition {
```

Comment:
Do we need this as a separate method?  Is there a way to have a default specialization so `assertAs` can have a base Error class that it throws if no specific one is specified?

Reply:
I guess I ask mostly because I don't love the name `assertAs` and would rather use `assert` for everything

Reply:
I tried to implement it as a single function but I couldn't find a simple way to support both overloads with TypeScript, so it was complaining about the arguments provided.
We can change the `assertAs` name with anything you'd find more suited though if you have any suggestion?

Reply:
I see, given what you say below I think this is fine as we won't actually be using `assertAs` everywhere throughout our code.  It'll just be used within the code that generates the `assert` object for each namespace's error class, then you'll use that assert object throughout the code.